In [69]:
#Importing pertinent libraries

import numpy as np
import pandas as pd
import warnings
ignore_warnings = True
import pyodbc

In [70]:
#Database connection
server = 'DESKTOP-59E71Q2\\SQLEXPRESS'
database = 'AdventureWorks2022'
username = 'sa'
password = 'Amaan123'
 
# Connection string
connection_string = f'DRIVER={{SQL Server}};SERVER={server};DATABASE={database};UID={username};PWD={password}'

In [71]:
try:
    conn = pyodbc.connect(connection_string)
    print("Connected Successfully")
except Exception as e:
    print(f'Error: {e}')

Connected Successfully


In [72]:
queryper = "SELECT * FROM [Person].[Person]"
queryemp = "SELECT * FROM [HumanResources].[Employee]"
querydepthist = "SELECT * FROM [HumanResources].[EmployeeDepartmentHistory]"

person_df = pd.read_sql(queryper, conn)
employee_df = pd.read_sql(queryemp, conn)
dept_history_df = pd.read_sql(querydepthist, conn)

C:\Users\Admin\AppData\Local\Temp\ipykernel_7224\3249092576.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  person_df = pd.read_sql(queryper, conn)
C:\Users\Admin\AppData\Local\Temp\ipykernel_7224\3249092576.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employee_df = pd.read_sql(queryemp, conn)
C:\Users\Admin\AppData\Local\Temp\ipykernel_7224\3249092576.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  dept_history_df = pd.read_sql(querydepthist, conn)


**Easy**

In [74]:
#1. Select BusinessEntityID, FirstName, and LastName columns from the Person DataFrame.

show_cols = person_df[['BusinessEntityID', 'FirstName', 'LastName']]
show_cols.head()

BusinessEntityID FirstName    LastName
0                 1       Ken     Sánchez
1                 2     Terri       Duffy
2                 3   Roberto  Tamburello
3                 4       Rob     Walters
4                 5      Gail    Erickson

In [75]:
#2. Where first name starst with 'A'

a_fname = person_df[person_df['FirstName'].str.startswith('A')]
a_fname.head()

BusinessEntityID PersonType  NameStyle Title FirstName MiddleName  \
32                 33         EM      False  None     Annik          O   
46                 47         EM      False  None    Andrew          R   
108               109         EM      False  None     Alice          O   
114               115         EM      False  None    Angela          W   
136               137         EM      False  None    Anibal          T   

      LastName Suffix  EmailPromotion AdditionalContactInfo  \
32       Stahl   None               0                  None   
46        Hill   None               1                  None   
108      Ciccu   None               1                  None   
114  Barbariol   None               2                  None   
136      Sousa   None               0                  None   

                                          Demographics  \
32   <IndividualSurvey xmlns="http://schemas.micros...   
46   <IndividualSurvey xmlns="http://schemas.micros...   
108  <IndividualSurvey xmlns="http://schemas.micros...   
114  <IndividualSurvey xmlns="http://schemas.micros...   
136  <IndividualSurvey xmlns="http://schemas.micros...   

                                  rowguid ModifiedDate  
32   BFA2BDC8-208A-447B-8A4E-72A215E9E134   2008-12-10  
46   B6439F4F-E57D-4293-B0C8-8919591055A3   2009-02-15  
108  B5836E66-4251-4D4D-8F49-8E6D9E3C9D64   2008-11-30  
114  8E6ADFD0-2FE0-4EB9-93F2-A04FD877F6DC   2009-01-13  
136  9B75A538-958C-4C91-987A-B9BCCE683E8A   2009-02-16

In [76]:
#3. Sort the Person DataFrame by LastName in ascending order
sort_lname = person_df.sort_values(by='LastName', ascending=True)
sort_lname.head()

BusinessEntityID PersonType  NameStyle Title  FirstName MiddleName  \
284                285         SP      False   Mr.       Syed          E   
291                293         SC      False   Ms.  Catherine         R.   
1364              2170         GC      False  None        Kim       None   
37                  38         EM      False  None        Kim          B   
292                295         SC      False   Ms.        Kim       None   

         LastName Suffix  EmailPromotion  \
284         Abbas   None               0   
291          Abel   None               1   
1364  Abercrombie   None               2   
37    Abercrombie   None               2   
292   Abercrombie   None               0   

                                  AdditionalContactInfo  \
284                                                None   
291   <AdditionalContactInfo xmlns="http://schemas.m...   
1364                                               None   
37                                                 None   
292   <AdditionalContactInfo xmlns="http://schemas.m...   

                                           Demographics  \
284   <IndividualSurvey xmlns="http://schemas.micros...   
291   <IndividualSurvey xmlns="http://schemas.micros...   
1364  <IndividualSurvey xmlns="http://schemas.micros...   
37    <IndividualSurvey xmlns="http://schemas.micros...   
292   <IndividualSurvey xmlns="http://schemas.micros...   

                                   rowguid            ModifiedDate  
284   FF284881-01C2-4C77-95A7-4DB96F59BB70 2013-03-07 00:00:00.000  
291   D54E0552-C226-4C22-AF3B-762CA854CDD3 2015-04-15 16:33:33.077  
1364  24F01B54-7A67-4B48-9ECC-72545D36C0AC 2009-11-29 00:00:00.000  
37    9A2163B3-2F4D-4F9A-91BD-07D326140F9C 2010-01-09 00:00:00.000  
292   F7CBDB48-0B44-470E-9F37-7060446FBFB9 2015-04-15 16:33:33.077

In [77]:
#4. Retrieve distinct FirstName values from the Person DataFrame.

distinct_fnames = person_df['FirstName'].unique()
dist_fnames = pd.DataFrame(distinct_fnames)  #Converting array returned by unique() function to df
dist_fnames.head()

0
0      Ken
1    Terri
2  Roberto
3      Rob
4     Gail

In [78]:
#5. Count the number of records in the Person DataFrame.
rcount = person_df.shape[0]
print('The number of records in Person are: ', rcount)

The number of records in Person are:  19972


In [79]:
#6. Filter records where LastName ends with 'son'
son_lname = person_df[person_df['LastName'].str.endswith('son')]
son_lname.head()

BusinessEntityID PersonType  NameStyle Title FirstName MiddleName  \
4                   5         EM      False   Ms.      Gail          A   
22                 23         EM      False  None      Mary          E   
42                 43         EM      False  None     Nancy          A   
48                 49         EM      False  None     Barry          K   
166               167         EM      False  None     David          N   

     LastName Suffix  EmailPromotion AdditionalContactInfo  \
4    Erickson   None               0                  None   
22     Gibson   None               0                  None   
42   Anderson   None               1                  None   
48    Johnson   None               0                  None   
166   Johnson   None               0                  None   

                                          Demographics  \
4    <IndividualSurvey xmlns="http://schemas.micros...   
22   <IndividualSurvey xmlns="http://schemas.micros...   
42   <IndividualSurvey xmlns="http://schemas.micros...   
48   <IndividualSurvey xmlns="http://schemas.micros...   
166  <IndividualSurvey xmlns="http://schemas.micros...   

                                  rowguid ModifiedDate  
4    F3A3F6B4-AE3B-430C-A754-9F2231BA6FEF   2007-12-30  
22   B10CF8EB-5589-4A81-B1BD-051C5759CE06   2009-01-05  
42   0681C8F9-D2EC-47D9-9BD4-751B14EAED50   2008-12-26  
48   CB28BE2E-DE93-4409-BACE-DF6042D04815   2013-11-29  
166  C2C7021B-3209-45F7-8560-B3132595F9DA   2008-11-25

In [80]:
#7. Select only the EmailAddress column from the Person DataFrame
# Loading email address table, since Person doesn't have email address column
emailquery = "SELECT * FROM [Person].[EmailAddress]"
email_df = pd.read_sql(emailquery,conn)

emails = email_df['EmailAddress']
emails.head()

C:\Users\Admin\AppData\Local\Temp\ipykernel_7224\1152943443.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  email_df = pd.read_sql(emailquery,conn)


0        ken0@adventure-works.com
1      terri0@adventure-works.com
2    roberto0@adventure-works.com
3        rob0@adventure-works.com
4       gail0@adventure-works.com
Name: EmailAddress, dtype: object

In [81]:
#8. Check for missing values in person_df

person_df.isna().sum()


BusinessEntityID             0
PersonType                   0
NameStyle                    0
Title                    18963
FirstName                    0
MiddleName                8499
LastName                     0
Suffix                   19919
EmailPromotion               0
AdditionalContactInfo    19962
Demographics                 0
rowguid                      0
ModifiedDate                 0
dtype: int64

Hence, in the person dataframe, Title, Middlename, Suffix and AdditionalContactInfo have 18963, 8499, 19919 and 19962 null values respectively

**Medium**

In [84]:
#9. Merge the Person and Employee DataFrames on BusinessEntityID using an inner join.
merged = pd.merge(person_df, employee_df, on='BusinessEntityID', how='inner')

merged.head()

BusinessEntityID PersonType  NameStyle Title FirstName MiddleName  \
0                 1         EM      False  None       Ken          J   
1                 2         EM      False  None     Terri        Lee   
2                 3         EM      False  None   Roberto       None   
3                 4         EM      False  None       Rob       None   
4                 5         EM      False   Ms.      Gail          A   

     LastName Suffix  EmailPromotion AdditionalContactInfo  ...   BirthDate  \
0     Sánchez   None               0                  None  ...  1969-01-29   
1       Duffy   None               1                  None  ...  1971-08-01   
2  Tamburello   None               0                  None  ...  1974-11-12   
3     Walters   None               0                  None  ...  1974-12-23   
4    Erickson   None               0                  None  ...  1952-09-27   

  MaritalStatus Gender    HireDate SalariedFlag VacationHours  SickLeaveHours  \
0             S      M  2009-01-14         True            99              69   
1             S      F  2008-01-31         True             1              20   
2             M      M  2007-11-11         True             2              21   
3             S      M  2007-12-05        False            48              80   
4             M      F  2008-01-06         True             5              22   

  CurrentFlag                             rowguid_y ModifiedDate_y  
0        True  F01251E5-96A3-448D-981E-0F99D789110D     2014-06-30  
1        True  45E8F437-670D-4409-93CB-F9424A40D6EE     2014-06-30  
2        True  9BBBFB2C-EFBB-4217-9AB7-F97689328841     2014-06-30  
3        True  59747955-87B8-443F-8ED4-F8AD3AFDF3A9     2014-06-30  
4        True  EC84AE09-F9B8-4A15-B4A9-6CCBAB919B08     2014-06-30  

[5 rows x 28 columns]

In [85]:
#10. Filter employees in the Employee DataFrame who have JobTitle containing 'Manager'.
managers = employee_df[employee_df['JobTitle'].str.contains('Manager')]
managers.head()

BusinessEntityID NationalIDNumber                   LoginID  \
2                  3        509647174  adventure-works\roberto0   
6                  7        134969118    adventure-works\dylan0   
9                 10        879342154  adventure-works\michael6   
15                16         24756624    adventure-works\david0   
25                26        277173473    adventure-works\peter0   

   OrganizationNode  OrganizationLevel                          JobTitle  \
2          b'Z\xc0'                2.0               Engineering Manager   
6          b'Z\xe1'                3.0  Research and Development Manager   
9         b'Z\xe1x'                4.0  Research and Development Manager   
15             b'h'                1.0                 Marketing Manager   
25         b'z\xc0'                2.0        Production Control Manager   

     BirthDate MaritalStatus Gender    HireDate  SalariedFlag  VacationHours  \
2   1974-11-12             M      M  2007-11-11          True              2   
6   1987-02-24             M      M  2009-02-08          True             61   
9   1984-11-30             M      M  2009-05-03          True             16   
15  1975-03-19             S      M  2007-12-20          True             40   
25  1982-11-03             M      M  2008-12-01          True             43   

    SickLeaveHours  CurrentFlag                               rowguid  \
2               21         True  9BBBFB2C-EFBB-4217-9AB7-F97689328841   
6               50         True  4F46DECA-EF01-41FD-9829-0ADAB368E431   
9               64         True  EAA43680-5571-40CB-AB1A-3BF68F04459E   
15              40         True  E87029AA-2CBA-4C03-B948-D83AF0313E28   
25              41         True  69D5D162-E817-45E7-9DEC-5D9B8310E7B1   

              ModifiedDate  
2  2014-06-30 00:00:00.000  
6  2014-06-30 00:00:00.000  
9  2014-06-30 00:00:00.000  
15 2014-06-30 00:00:00.000  
25 2014-12-26 09:17:08.637

In [86]:
#11. Extract the year from the HireDate column in the Employee DataFrame and add it as a new column.

employee_df['HireDate'] = pd.to_datetime(employee_df['HireDate'])
employee_df['HYear'] = employee_df['HireDate'].dt.year

employee_df.head()

BusinessEntityID NationalIDNumber                   LoginID  \
0                 1        295847284      adventure-works\ken0   
1                 2        245797967    adventure-works\terri0   
2                 3        509647174  adventure-works\roberto0   
3                 4        112457891      adventure-works\rob0   
4                 5        695256908     adventure-works\gail0   

  OrganizationNode  OrganizationLevel                       JobTitle  \
0             None                NaN        Chief Executive Officer   
1             b'X'                1.0  Vice President of Engineering   
2         b'Z\xc0'                2.0            Engineering Manager   
3         b'Z\xd6'                3.0           Senior Tool Designer   
4         b'Z\xda'                3.0                Design Engineer   

    BirthDate MaritalStatus Gender   HireDate  SalariedFlag  VacationHours  \
0  1969-01-29             S      M 2009-01-14          True             99   
1  1971-08-01             S      F 2008-01-31          True              1   
2  1974-11-12             M      M 2007-11-11          True              2   
3  1974-12-23             S      M 2007-12-05         False             48   
4  1952-09-27             M      F 2008-01-06          True              5   

   SickLeaveHours  CurrentFlag                               rowguid  \
0              69         True  F01251E5-96A3-448D-981E-0F99D789110D   
1              20         True  45E8F437-670D-4409-93CB-F9424A40D6EE   
2              21         True  9BBBFB2C-EFBB-4217-9AB7-F97689328841   
3              80         True  59747955-87B8-443F-8ED4-F8AD3AFDF3A9   
4              22         True  EC84AE09-F9B8-4A15-B4A9-6CCBAB919B08   

  ModifiedDate  HYear  
0   2014-06-30   2009  
1   2014-06-30   2008  
2   2014-06-30   2007  
3   2014-06-30   2007  
4   2014-06-30   2008

In [87]:
#12. Create an intermediate DataFrame to calculate the number of employees in each department.

employees_by_dept = dept_history_df.groupby('DepartmentID')['BusinessEntityID'].count().reset_index(name='NumEmps')
employees_by_dept

DepartmentID  NumEmps
0              1        7
1              2        4
2              3       18
3              4       10
4              5       13
5              6        4
6              7      180
7              8        6
8              9        6
9             10       11
10            11       10
11            12        5
12            13        7
13            14        7
14            15        6
15            16        2

In [88]:
#13 Group by JobTitle and count the number of employees in each group

employees_by_job = employee_df.groupby('JobTitle')['BusinessEntityID'].count().reset_index(name = 'EmpsbyJob')
sorted_empjob = employees_by_job.sort_values('EmpsbyJob', ascending = False)
sorted_empjob

JobTitle  EmpsbyJob
44  Production Technician - WC40         26
47  Production Technician - WC60         26
46  Production Technician - WC50         26
43  Production Technician - WC30         25
42  Production Technician - WC20         22
..                           ...        ...
26             Marketing Manager          1
28              Master Scheduler          1
30               Network Manager          1
31  North American Sales Manager          1
66       Vice President of Sales          1

[67 rows x 2 columns]

In [120]:
#14. Filter employees hired in 2020
h_2020 = employee_df[employee_df['HYear'] == 2020]
h_2020.head()

employee_df['HYear'].max()  #2013

#Since 2013, is the latest HireYear in the dataframe, the data will be blank for 2020. We can check for other
# years by substituting the value in the above code by another year.

h_2011 = employee_df[employee_df['HYear'] == 2011]
h_2011.head()


BusinessEntityID NationalIDNumber                  LoginID  \
14                 15         56920285  adventure-works\sharon0   
17                 18        222969461    adventure-works\john5   
18                 19         52541318    adventure-works\mary2   
19                 20        323403273  adventure-works\wanida0   
249               250        895209680  adventure-works\sheela0   

    OrganizationNode  OrganizationLevel              JobTitle   BirthDate  \
14          b'Z\xe7'                3.0       Design Engineer  1961-05-02   
17             b'k@'                2.0  Marketing Specialist  1978-03-06   
18          b'k\xc0'                2.0   Marketing Assistant  1978-01-29   
19             b'l '                2.0   Marketing Assistant  1975-03-17   
249      b'\x85\xeb'                3.0    Purchasing Manager  1978-02-10   

    MaritalStatus Gender   HireDate  SalariedFlag  VacationHours  \
14              M      F 2011-01-18          True              4   
17              S      M 2011-02-07         False             48   
18              S      F 2011-02-14         False             43   
19              M      F 2011-01-07         False             41   
249             S      F 2011-02-25          True             49   

     SickLeaveHours  CurrentFlag                               rowguid  \
14               22         True  54F2FDC0-87C4-4065-A7A8-9AC8EA624235   
17               44         True  64730415-1F58-4E5B-8FA8-5E4DAEBA53B4   
18               41         True  1F6DA901-C7F7-48A8-8EEF-D81868D72B52   
19               40         True  43CCA446-DA1C-454C-8530-873AD2923E1B   
249              44         True  7D6B0155-0191-4599-8394-378EBDE63C9A   

    ModifiedDate  HYear  
14    2014-06-30   2011  
17    2014-06-30   2011  
18    2014-06-30   2011  
19    2014-06-30   2011  
249   2014-06-30   2011

In [130]:
#15. Calculate the average years of experience for employees in the Employee DataFrame.

c_year = pd.Timestamp.now().year
employee_df['ExperienceYears'] = c_year - employee_df['HYear']

avg_exp = employee_df['ExperienceYears'].mean()
print("Average years of experience for employees is: ", avg_exp.round(2), " years")

Average years of experience for employees is:  14.98  years


In [132]:
#16. Rename the columns in the Person DataFrame to have lowercase names.
person_df_copy = person_df

person_df_copy.columns = person_df_copy.columns.str.lower()
person_df_copy.head()


businessentityid persontype  namestyle title firstname middlename  \
0                 1         EM      False  None       Ken          J   
1                 2         EM      False  None     Terri        Lee   
2                 3         EM      False  None   Roberto       None   
3                 4         EM      False  None       Rob       None   
4                 5         EM      False   Ms.      Gail          A   

     lastname suffix  emailpromotion additionalcontactinfo  \
0     Sánchez   None               0                  None   
1       Duffy   None               1                  None   
2  Tamburello   None               0                  None   
3     Walters   None               0                  None   
4    Erickson   None               0                  None   

                                        demographics  \
0  <IndividualSurvey xmlns="http://schemas.micros...   
1  <IndividualSurvey xmlns="http://schemas.micros...   
2  <IndividualSurvey xmlns="http://schemas.micros...   
3  <IndividualSurvey xmlns="http://schemas.micros...   
4  <IndividualSurvey xmlns="http://schemas.micros...   

                                rowguid modifieddate  
0  92C4279F-1207-48A3-8448-4636514EB7E2   2009-01-07  
1  D8763459-8AA8-47CC-AFF7-C9079AF79033   2008-01-24  
2  E1A2555E-0828-434B-A33B-6F38136A37DE   2007-11-04  
3  F2D7CE06-38B3-4357-805B-F4B6B71C01FF   2007-11-28  
4  F3A3F6B4-AE3B-430C-A754-9F2231BA6FEF   2007-12-30

**Hard**

In [157]:
#17. Write a recursive function to find all employees reporting under a specific manager in the Employee DataFrame.
# (Manager data missing, hence subordinates calculated by OrganizationLevel, where there is no single level for single person, hence we can
# fetch lower level employees without recursion).


def find_subs(employee_df, employee_id):
    upper_level = employee_df[employee_df['BusinessEntityID'] == employee_id]['OrganizationLevel'].values[0]
    subs = employee_df[employee_df['OrganizationLevel'] < upper_level]
    subs = subs.sort_values('OrganizationLevel', ascending = False)
    return subs

employee_df[employee_df['BusinessEntityID'] == 121]['OrganizationLevel'] #3.0
find_subs(employee_df, 121)

BusinessEntityID NationalIDNumber                     LoginID  \
286               287        982310417        adventure-works\amy0   
271               272        525932996    adventure-works\janaina0   
269               270        643805155   adventure-works\françois0   
268               269        671089628        adventure-works\dan0   
267               268        314747499     adventure-works\ramesh0   
266               267         58317344      adventure-works\karen1   
263               264        858323870  adventure-works\stephanie0   
261               262        231203233      adventure-works\david5   
248               249        121491555      adventure-works\wendy0   
240               241            30845      adventure-works\david6   
234               235        535145551      adventure-works\paula0   
2                   3        509647174    adventure-works\roberto0   
226               227        141165819       adventure-works\gary1   
221               222        685233686     adventure-works\ascott0   
210               211        398223854      adventure-works\hazem0   
25                 26        277173473      adventure-works\peter0   
273               274        502097814    adventure-works\stephen0   
23                 24         72636981       adventure-works\jill0   
22                 23        767955365       adventure-works\mary0   
21                 22         95958330     adventure-works\sariya0   
20                 21        243322160      adventure-works\terry0   
19                 20        323403273     adventure-works\wanida0   
18                 19         52541318       adventure-works\mary2   
17                 18        222969461       adventure-works\john5   
16                 17        253022876      adventure-works\kevin0   
284               285        481044938       adventure-works\syed0   
270               271        929666391        adventure-works\dan1   
272               273        112432117      adventure-works\brian3   
1                   2        245797967      adventure-works\terri0   
262               263        441044382       adventure-works\jean0   
24                 25        519899904      adventure-works\james1   
15                 16         24756624      adventure-works\david0   
233               234        184188301      adventure-works\laura1   

    OrganizationNode  OrganizationLevel  \
286      b'\x95\xe0'                2.0   
271         b'\x8ep'                2.0   
269         b'\x8e0'                2.0   
268      b'\x8e\x10'                2.0   
267      b'\x8d\xe0'                2.0   
266      b'\x8d\xa0'                2.0   
263         b'\x8d`'                2.0   
261      b'\x86\x10'                2.0   
248      b'\x85\xe0'                2.0   
240      b'\x85\xa0'                2.0   
234         b'\x85`'                2.0   
2           b'Z\xc0'                2.0   
226            b'| '                2.0   
221         b'{\xc0'                2.0   
210            b'{@'                2.0   
25          b'z\xc0'                2.0   
273         b'\x95`'                2.0   
23          b'm\x10'                2.0   
22          b'l\xe0'                2.0   
21          b'l\xa0'                2.0   
20             b'l`'                2.0   
19             b'l '                2.0   
18          b'k\xc0'                2.0   
17             b'k@'                2.0   
16          b'j\xc0'                2.0   
284      b'\x95\xa0'                2.0   
270         b'\x8eP'                2.0   
272          b'\x94'                1.0   
1               b'X'                1.0   
262          b'\x8c'                1.0   
24              b'x'                1.0   
15              b'h'                1.0   
233          b'\x84'                1.0   

                                     JobTitle   BirthDate MaritalStatus  \
286                    European Sales Manager  1957-09-20             M   
271        

In [165]:
#18. Rank employees by HireDate
employee_df['HireDateRank'] = employee_df['HireDate'].rank()

employee_df[['BusinessEntityID', 'HireDate', 'HireDateRank']].sort_values('HireDateRank')


BusinessEntityID   HireDate  HireDateRank
27                 28 2006-06-30           1.0
16                 17 2007-01-26           2.0
2                   3 2007-11-11           3.0
3                   4 2007-12-05           4.0
11                 12 2007-12-11           5.0
..                ...        ...           ...
289               290 2012-05-30         285.5
283               284 2012-09-30         287.0
284               285 2013-03-14         288.0
285               286 2013-05-30         289.5
287               288 2013-05-30         289.5

[290 rows x 3 columns]

In [185]:
#19. Define a function to retrieve employee details based on DepartmentID and HireDate.

def get_details(employee_df, dept_history_df, department_id, hire_date):
    dept_emps = dept_history_df[dept_history_df['DepartmentID'] == department_id]
    merged = pd.merge(employee_df, dept_emps, on='BusinessEntityID', how='inner')
    details_df = merged[merged['HireDate'] == hire_date]
    return details_df


employee_details = get_details(employee_df, dept_history_df, department_id=3, hire_date='2013-05-30')
employee_details


BusinessEntityID NationalIDNumber                  LoginID  \
13               286        758596752    adventure-works\lynn0   
15               288        954276278  adventure-works\rachel0   

   OrganizationNode  OrganizationLevel              JobTitle   BirthDate  \
13      b'\x95\xab'                3.0  Sales Representative  1977-02-14   
15      b'\x95\xeb'                3.0  Sales Representative  1975-07-09   

   MaritalStatus Gender   HireDate  ...                               rowguid  \
13             S      F 2013-05-30  ...  4A9A8407-A680-4A6B-8D03-511CB58F9A8A   
15             S      F 2013-05-30  ...  B9BF7741-E0CA-4F37-ACDE-A4F78C6D03E9   

    ModifiedDate_x  HYear  ExperienceYears HireDateRank DepartmentID  ShiftID  \
13      2014-06-30   2013               11        289.5            3        1   
15      2014-06-30   2013               11        289.5            3        1   

     StartDate  EndDate  ModifiedDate_y  
13  2013-05-30     None      2013-05-29  
15  2013-05-30     None      2013-05-29  

[2 rows x 24 columns]

In [191]:
#20. Find employees hired in the last 30 days
#Data is old, hence taking sample date from which 30 day period is calculated - current date can be used by pd.Timestamp.now()

thirty_days_ago = pd.to_datetime('2013-06-01') - pd.DateOffset(days=30)
hired = employee_df[employee_df['HireDate'] >= thirty_days_ago]

hired

BusinessEntityID NationalIDNumber                  LoginID  \
285               286        758596752    adventure-works\lynn0   
287               288        954276278  adventure-works\rachel0   

    OrganizationNode  OrganizationLevel              JobTitle   BirthDate  \
285      b'\x95\xab'                3.0  Sales Representative  1977-02-14   
287      b'\x95\xeb'                3.0  Sales Representative  1975-07-09   

    MaritalStatus Gender   HireDate  SalariedFlag  VacationHours  \
285             S      F 2013-05-30          True             36   
287             S      F 2013-05-30          True             35   

     SickLeaveHours  CurrentFlag                               rowguid  \
285              38         True  4A9A8407-A680-4A6B-8D03-511CB58F9A8A   
287              37         True  B9BF7741-E0CA-4F37-ACDE-A4F78C6D03E9   

    ModifiedDate  HYear  ExperienceYears  HireDateRank  
285   2014-06-30   2013               11         289.5  
287   2014-06-30   2013               11         289.5

In [219]:
#21. Calculate the retention rate of employees over the past 5 years - no EndDate available in employee, hence using dept_history_df
five_years_ago = pd.to_datetime('2013-12-31') - pd.DateOffset(years = 5)

dept_history_df['StartDate'] = pd.to_datetime(dept_history_df['StartDate'])
dept_history_df['EndDate'] = pd.to_datetime(dept_history_df['EndDate'])
                                            
hired_5_years_ago = dept_history_df[dept_history_df['StartDate'] <= five_years_ago]
retained_employees = hired_5_years_ago[hired_5_years_ago['EndDate'].isnull()]

retention_rate = len(retained_employees) / len(hired_5_years_ago) * 100

print('Retention Rate is: ',retention_rate)

Retention Rate is:  97.53086419753086


In [223]:
#22. Identify employees who have changed departments more than twice

dept_changes = dept_history_df.groupby('BusinessEntityID').size()

multiple_changes = dept_changes[dept_changes > 2].reset_index(name='ChangeCount')
multiple_changes


BusinessEntityID  ChangeCount
0               250            3

In [243]:
#23. Analyze the distribution of employees across different job titles

distribution = employee_df['JobTitle'].value_counts()
distribution

JobTitle
Production Technician - WC60           26
Production Technician - WC40           26
Production Technician - WC50           26
Production Technician - WC30           25
Production Technician - WC20           22
                                       ..
Facilities Manager                      1
Maintenance Supervisor                  1
Facilities Administrative Assistant     1
Chief Financial Officer                 1
European Sales Manager                  1
Name: count, Length: 67, dtype: int64

In [247]:
#24. Determine the maximum, minimum, and average hire dates for each department

dept_hire_stats = dept_history_df.groupby('DepartmentID').agg(
    MaxHireDate=('StartDate', 'max'),
    MinHireDate=('StartDate', 'min'),
    AvgHireDate=('StartDate', 'mean')
)

dept_hire_stats

MaxHireDate MinHireDate                   AvgHireDate
DepartmentID                                                      
1             2011-01-18  2007-11-11 2008-11-08 10:17:08.571428608
2             2010-12-23  2007-12-11 2010-01-24 00:00:00.000000000
3             2013-05-30  2011-01-04 2011-12-05 21:20:00.000000000
4             2011-02-25  2007-01-26 2009-09-19 21:36:00.000000000
5             2012-07-15  2007-12-20 2010-01-31 01:50:46.153846272
6             2009-05-03  2008-12-29 2009-02-13 00:00:00.000000000
7             2010-03-10  2006-06-30 2009-02-26 18:08:00.000000000
8             2011-09-01  2008-12-01 2009-06-25 04:00:00.000000000
9             2009-02-25  2008-12-06 2009-01-07 12:00:00.000000000
10            2009-03-08  2008-12-18 2009-01-27 10:54:32.727272704
11            2009-02-23  2008-12-04 2009-01-16 12:00:00.000000000
12            2009-03-06  2008-12-16 2009-01-23 04:48:00.000000000
13            2011-07-31  2008-12-09 2009-07-25 03:25:42.857142784
14            2010-03-07  2008-12-14 2009-11-30 00:00:00.000000000
15            2009-02-26  2008-12-07 2009-01-04 08:00:00.000000000
16            2013-11-14  2009-01-14 2011-06-15 12:00:00.000000000